### ISSUE TITLE
Nginx proxy config in a folder

### ISSUE THREAD

**Issue by RNCTX** [babelouest/glewlwyd#152]
I'd like to use this app on a shared subdomain with other apps, each residing in a folder that Nginx proxies to based on the directory in the URL.

I've tried the following:

**Nginx config:**

```
 # Glewlwyd
        location ~/authsrv(.*?)$ {
            proxy_set_header Authorization $http_authorization;
            proxy_set_header Host $host;
            proxy_pass_header  Authorization;
            proxy_pass         http://127.0.0.1:4593$1;
        }
```

**Glewlwyd.conf:**

```
# port to open for remote commands
port=4593

# bind to IPV4 address
bind_address="127.0.0.1"

# external url to access to this instance
external_url="https://site.url"

# login url relative to external url
login_url="login.html"

# url prefix
api_prefix="api"

# path to static files for /webapp url
static_files_path="/usr/share/glewlwyd/webapp/"
```

**webapp/config.json:**

```
{
  "GlewlwydUrl": "https://site.url/authsrv/",
  "ProfileUrl": "profile.html",
  "AdminUrl": "index.html",
  "LoginUrl": "login.html",
  "CallbackPage": "callback.html",
```

I would think this would be pretty transparent to Glewlwyd, since Nginx is rewriting the URLs as they come through. With this I can log in, log out, and get pages to load but callback redirects don't work. I get a 400 error about scope with the following:

![image](https://user-images.githubusercontent.com/13775820/87227502-10d72280-c361-11ea-8744-b1af8587d4c3.png)

Is there something else I need to do, or a better way, to accomplish Nginx as a reverse proxy aimed at Glewlwyd in a subfolder?

**Comment by babelouest** [comment]
In the endpoint `GET` `/auth/scheme`, the API returns an error 400 if the scope parameter is empty. My intuition is that your nginx proxy configuration doesn't pass the query parameters to Glewlwyd.

**Comment by RNCTX** [comment]
That was my suspicion as well. I found the other report awhile back about having to manually add the authorization header. Will work on the Nginx docs a bit and report back for others when I get it working, thanks!

**Comment by babelouest** [comment]
> Will work on the Nginx docs a bit and report back for others when I get it working, thanks!

Thanks!

If you're ok, open a pull request to update the new section `Reverse proxy configuration` in the [Install documentation](https://github.com/babelouest/glewlwyd/blob/master/docs/INSTALL.md#reverse-proxy-configuration) with a new paragraph for Nginx proxy configuration.

**Comment by RNCTX** [comment]
Will do!

The answer to this is that there is no answer with location matching, because Nginx will only pass URL queries and params via rewrite rules, so here's a working config for Nginx in a subfolder:

```
 # Glewlwyd
        location ~ ^/authsrv {
            rewrite ^\/authsrv(.*) /$1 break;
            proxy_set_header Authorization $http_authorization;
            proxy_set_header Host $host;
            proxy_pass_header  Authorization;
            proxy_pass         http://127.0.0.1:4593;
        }
```

This will listen on the subdirectory https://site.url/authsrv but pass all requests to the backend at the root URL the backend expects.  Only the webapp config needs to be changed to reflect the /authsrv directory, for redirect purposes.

I'll submit a docs PR later.

**Comment by babelouest** [comment]
Nice!

Why must the header `Authorization` be specifically passed in the proxy config? I assume all the other headers are forwarded. Is there an exception for the `Authorization` header?

I also assume this config goes for https reverse proxy as well?

**Comment by babelouest** [comment]
Also, I write it down here to not forget about it, but I'd like to add the client ssl certificate proxy config in the documentation. For what I've seen, I have to add the following settings:

```
ssl_client_certificate /etc/ssl/certs/ca.crt;
ssl_verify_client on | optional | optional_no_ca
proxy_set_header        SSL_CLIENT_CERT         $client_cert;
```

**Comment by RNCTX** [comment]
Yes, I tried to disable the certificate settings in the glewdwyd.conf but that got me an error, if I remember correctly, so it seems that one cannot comment them out even if SSL is disabled?

> Why must the header Authorization be specifically passed in the proxy config? I assume all the other headers are forwarded. Is there an exception for the Authorization header?

Because Nginx ignores any headers not present in the *initial* request of a session.  Because an unauthorized user has no Authentication header until they log in, Nginx will not pass it in subsequent requests unless it is explicitly declared for the first request to the proxy app/service.

> I also assume this config goes for https reverse proxy as well?

It should, I am terminating SSL at Nginx personally, and if someone is not terminating SSL at the proxy the `ssl_stream_preread` module is a much easier way to pass encrypted traffic through Nginx (it allows Nginx  to sniff out the requested address but pass the encrypted traffic unmodified, so all the user has to do is specify their backend host(s) in their load balancing setup).

**Comment by babelouest** [comment]
> Because Nginx ignores any headers not present in the initial request of a session

Not really stateless but ok

> I presume that in the glewlwyd web app an unauthorized user has no Authentication header until they log in

Well, technically, Glewlwyd's webapp uses session cookies to authenticate users. The `Authorization` header is used in the oidc and oauth2 plugins.

> It should, but anyone wanting to pass SSL through Nginx as a proxy should not be bothering with certificates at the proxy,

I'm not sure we're talking about the same thing, I was talking about having a Nginx as a reverse proxy, listening on `https://` and forwarding to a glewlwyd on `http://`.

Thanks for the pull request! I think this bug can be closed now, don't you?

**Comment by RNCTX** [comment]
Yes, this is all working fine for me.  I'll post back in this thread about what I find with the `Authorization` header.  I found your project after failing to get Keycloak to work for a couple of days with an API backend that I am trying to evaluate, which simply would not authenticate properly no matter what I tried (using openid).

Yours is much lighter which is also a plus for a small server without much RAM, but after setting up both I wonder if Nginx  not passing `Authorization` with proxy requests was the issue all along. I'm going by the report in #32 about the need to add it.
